### ISSUE TITLE
Fix S4158: "Empty collections" raises false positive when analyzed method has too many branches

### ISSUE THREAD

**Issue by nnpcYvIVl** [SonarSource/sonar-csharp#767]
A false positive S4158 is thrown on the line: 
```csharp 
datagridview[columnname, datagridviewrow.Index].Style.BackColor = columnnamecolordictionary[columnname];
```

```csharp
public static void ColorRow(UserControl usercontrol, DataGridViewRow datagridviewrow, List<string> columnnamelist)
{
    string name = usercontrol.Name.Replace("UserControl", string.Empty);
    DataGridView datagridview = datagridviewrow.DataGridView;

    Dictionary<string, Color> columnnamecolordictionary = new Dictionary<string, Color>();

    for (int i = 0; i < datagridviewrow.DataGridView.Columns.Count; i++)
    {
        foreach (string columnname in columnnamelist)
        {
            if ((columnname == name || columnname == "*") && columnname == datagridviewrow.DataGridView.Columns[i].Name && datagridviewrow.Cells[i].Value != DBNull.Value && !string.IsNullOrWhiteSpace(datagridviewrow.Cells[i].Value.ToString()))
            {
                if (columnname == "=")
                {
                    // do nothing
                }

                columnnamecolordictionary.Add(columnname, Color.AliceBlue);
            }
        }
    }

    foreach (string columnname in columnnamecolordictionary.Keys)
    {
        if (datagridview.Columns.Contains(columnname))
        {
            datagridview[columnname, datagridviewrow.Index].Style.BackColor = columnnamecolordictionary[columnname];
        }
    }
}
```

**Comment by valhristov** [comment]
The false positive is caused because of our limited support for loops - the rule cannot detect that you are filling the dictionary with items a few lines above, hence it raises an issue.

**Comment by valhristov** [comment]
Apparently I diagnosed the problem wrong, it is caused because the method has rather many branches and our dataflow analysis engine reaches the limit of explorations count (1000). We would like to avoid increasing the limit for now because it could have performance impact. The rule will currently not report if it detects that the exploded graph exploration did not finish correctly (exactly as the condition evaluates to constant rule).
